### PR TITLE
UI adjustments (select label by arrow key, save/discard/cancel dialog, arrow labels.txt)

### PR DESCRIPTION
**Pull request by jbvo**
Some small UI adjustments that I implemented over the past few days while using labelme.

1. When closing or switching to another file, the dialog now asks to `save changes` aside of `discard changes` and `cancel`, more similar to other applications.
2. The label dialog preselects the previously selected label or the label that is under edit because of clicking in the main window label list. Also it allows to change the selected label of the list by using the arrow keys.
3. Extended loading labels from file, filtering empty line labels or clarity.

### CODE REVIEW

File: labelme/app.py [hunk 1]
```diff
@@ -794,7 +796,6 @@ def loadFile(self, filename=None):
             self.addRecentFile(self.filename)
             self.toggleActions(True)
             self.status("Loaded %s" % os.path.basename(str(filename)))
-            self.setWindowTitle('{} - {}'.format(__appname__, os.path.basename(str(filename))))
```
Comment (wkentaro) [inline labelme/app.py]: Why this change is necessary?

File: labelme/labelDialog.py [hunk 2]
```diff
@@ -89,11 +89,20 @@ def postProcess(self):
...
         self.edit.setFocus(Qt.PopupFocusReason)
         if move:
             self.move(QCursor.pos())
         return self.edit.text() if self.exec_() else None
 
+    def keyPressEvent(self, e):
+        if e.key() in [Qt.Key_Up, Qt.Key_Down]:
+            self.labelList.keyPressEvent(e)
+        else:
+            super().keyPressEvent(e)
```
Comment (wkentaro) [inline labelme/labelDialog.py]: The label changes using arrow keys without this change in my env.

File: labelme/app.py (hunk 1 above)
Comment (jbvo) [reply labelme/app.py]: Since I was added a `setWindowTitle` call to the `setDirty` method, I thought it would be clearer if this specific `setWindowTitle` call moved along with it from `loadFile` to `setClean`.

At this moment it won't make a difference if the call is in `loadFile` or in `setClean`, but if at some point an undo function would be implemented if should be in `setClean` anyway to remove the asterisk from the window title.

File: labelme/labelDialog.py [hunk 3]
```diff
@@ -90,19 +90,15 @@ def postProcess(self):
...
+        items = self.labelList.findItems(text, Qt.MatchFixedString)
+        if items:
+            assert len(items) == 1
+            self.labelList.setCurrentItem(items[0])
         self.edit.setFocus(Qt.PopupFocusReason)
         if move:
             self.move(QCursor.pos())
         return self.edit.text() if self.exec_() else None
-
-    def keyPressEvent(self, e):
```
Comment (jbvo) [inline labelme/labelDialog.py]: Why did you remove the keyPressEvent?

Comment (jbvo) [inline labelme/labelDialog.py]: Why did you remove the `keyPressEvent`?

Comment (wkentaro) [reply labelme/labelDialog.py]: Do you really need this?
The label changes using arrow keys without this change in my env. It is only on macOS..?

File: labelme/app.py (hunk 1 above)
Comment (wkentaro) [reply labelme/app.py]: Ok I see.

File: labelme/labelDialog.py (hunk 3 above)
Comment (wkentaro) [reply labelme/labelDialog.py]: Could you re-check this?

Comment (jbvo) [reply labelme/labelDialog.py]: I'm on Ubuntu 16.04. In my case arrow keys only work if the listwidget has the focus, but we set the focus to the editbox. This keypress event allows moving the listwidget selection without first moving the focus to the listwidget.

Comment (jbvo) [reply labelme/labelDialog.py]: I've double checked it and without the keypress event I have to either press TAB 3 times or click on the listwidget before I can use the arrow keys for selecting a label from the list.

File: labelme/labelDialog.py [hunk 4]
```diff
@@ -90,19 +90,15 @@ def postProcess(self):
             self.edit.setText(self.edit.text().trimmed())
 
     def popUp(self, text=None, move=True):
-        if isinstance(text, str):
-            self.edit.setText(text)
-            self.edit.setSelection(0, len(text))
-            i = self.labelList.findItems(text, Qt.MatchFixedString)
-            if len(i) == 1:
-                self.labelList.setCurrentItem(i[0])
+        if text is None:
```
Comment (jbvo) [inline labelme/labelDialog.py]: The reason I had put the complete code block inside the if was for the case of labeling many instances of the same label in 1 image. In that case it is convenient to not have to enter the label name again, but instead still having it from the previous label action.

If you prefer to not keep the previous label in the editbox, I would also suggest to add the following block below:

```py
if items:
    assert len(items) == 1
    self.labelList.setCurrentItem(items[0])
else:
    self.labelList.clearSelection()
```

File: labelme/labelDialog.py (hunk 3 above)
Comment (wkentaro) [reply labelme/labelDialog.py]: I understood. But the implementation did not work on macOS, so I implemented it with another approach.

File: labelme/labelDialog.py (hunk 4 above)
Comment (wkentaro) [reply labelme/labelDialog.py]: I think keeping the previous label is a good idea.
However, how can it be achieved?
Where do you keep the previous label?

Comment (wkentaro) [reply labelme/labelDialog.py]: Ah, now I understand. And I found it is necessary to ensure `text=None`.

**Comment by wkentaro** [comment]
@jbvo Could you check and review my commits?

File: labelme/labelDialog.py (hunk 3 above)
Comment (jbvo) [reply labelme/labelDialog.py]: For my understanding of python and pyQt: did the up and down key not work at all? Were the escape and enter key working for accepting and rejecting?

One thing I noticed when checking the differences is the call `else` case, my initial implementation:
```py
super().keyPressEvent(e)
```
your implementation:
```py
super(LabelQLineEdit, self).keyPressEvent(e)
```
Your implementation is more compatible as far as I understand from python docs, as `super()` was introduced in python 3 for convenience.

**Comment by jbvo** [comment]
I tested the adjusted label dialog and it works. I've put a question about the `keyPressEvent`, but that's not very critical, so merging would be fine for me as is.

File: labelme/labelDialog.py (hunk 3 above)
Comment (wkentaro) [reply labelme/labelDialog.py]: > For my understanding of python and pyQt: did the up and down key not work at all? Were the escape and enter key working for accepting and rejecting?

Hmm, I'm not sure about your question, but all keys, up, down, escape, enter key work well;
up and down keys are for selecting label from list, escape and enter keys are for accepting or rejecting.

Comment (wkentaro) [reply labelme/labelDialog.py]: > Your implementation is more compatible as far as I understand from python docs, as super() was introduced in python 3 for convenience.

Yes, my implementation is for compatibility on python2.
